src/ray/raylet/worker_pool.cc

```diff
@@ -184,9 +198,12 @@ Process WorkerPool::StartWorkerProcess(
   // Here we consider both task workers and I/O workers.
   if (starting_workers >= maximum_startup_concurrency_) {
     // Workers have been started, but not registered. Force start disabled -- returning.
-    RAY_LOG(DEBUG) << "Worker not started, " << starting_workers
-                   << " workers of language type " << static_cast<int>(language)
-                   << " pending registration";
+    std::ostringstream oss;
+    oss << "Worker not started, " << starting_workers << " workers of language type "
+        << static_cast<int>(language) << " pending registration";
+    RAY_LOG(DEBUG) << oss.str();
+    PopWorkerCallbackExecution(callback, nullptr,
```

Comment:
Why not wait to call the callback until the starting workers register?

Reply:
Yep, when the starting workers number greater than maximum_startup_concurrency_, we will not start an new worker process. We should re-dispatch this task.

Reply:
Let me see how to test it.

Reply:
Hmm I guess I thought that now that the callback is async, we should just wait until a worker was available. Right now, the logic will loop:
1. Trigger dispatch/PopWorker.
2. If startup concurrency maxed out, callback with empty worker.
But we could also avoid looping if we did this:
1. Trigger dispatch/PopWorker.
2. If startup concurrency maxed out, wait until a worker is pushed or a worker can be started, then callback.

I don't think it's a good idea to do it in this PR, but it would be good to do it in the future. Can you add as a TODO?

Reply:
Sure, Let me add this TODO.